File: api/renderer/python/indigo_renderer.py
#
# Copyright (C) 2010-2011 GGA Software Services LLC
# 
# This file is part of Indigo toolkit.
# 
# This file may be distributed and/or modified under the terms of the
# GNU General Public License version 3 as published by the Free Software
# Foundation and appearing in the file LICENSE.GPL included in the
# packaging of this file.
# 
# This file is provided AS IS with NO WARRANTY OF ANY KIND, INCLUDING THE
# WARRANTY OF DESIGN, MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.

import indigo
from indigo import *

class IndigoRenderer:

  def __init__ (self, indigo):
    self.indigo = indigo

    if os.name == 'posix' and not platform.mac_ver()[0]:
      self._lib = CDLL(indigo.dllpath + "/libindigo-renderer.so")
    elif os.name == 'nt':
      self._lib = CDLL(indigo.dllpath + "\indigo-renderer.dll")
    elif platform.mac_ver()[0]:
      self._lib = CDLL(indigo.dllpath + "/libindigo-renderer.dylib")
    else:
      raise IndigoException("unsupported OS: " + os.name)

    self._lib.indigoRender.restype = c_int
    self._lib.indigoRender.argtypes = [c_int, c_int]
    self._lib.indigoRenderToFile.restype = c_int
    self._lib.indigoRenderToFile.argtypes = [c_int, c_char_p]
    self._lib.indigoRenderGrid.restype = c_int
    self._lib.indigoRenderGrid.argtypes = [c_int, POINTER(c_int), c_int, c_int]
    self._lib.indigoRenderGridToFile.restype = c_int
    self._lib.indigoRenderGridToFile.argtypes = [c_int, POINTER(c_int), c_int, c_char_p]
    self._lib.indigoRenderReset.restype = c_int
    self._lib.indigoRenderReset.argtypes = [c_int]
    
  def renderToBuffer (self, obj):
    self.indigo._setSID()
    wb = self.indigo.writeBuffer()
    self.indigo._checkResult(self._lib.indigoRender(obj.id, wb.id))
    return wb.toBuffer()
  
  def renderToFile (self, obj, filename):
    self.indigo._setSID()
    self.indigo._checkResult(self._lib.indigoRenderToFile(obj.id, filename))

  def renderGridToFile (self, objects, refatoms, ncolumns, filename):
    self.indigo._setSID()
    arr = None
    if refatoms:
      if len(refatoms) != objects.size():
        raise IndigoException("renderGridToFile(): refatoms[] size must be equal to the number of objects")
      arr = (c_int * len(refatoms))()
      for i in xrange(len(refatoms)):
        arr[i] = refatoms[i]
    self.indigo._checkResult(
      self._lib.indigoRenderGridToFile(objects.id, arr, ncolumns, filename))

  def renderGridToBuffer (self, objects, refatoms, ncolumns):
    self.indigo._setSID()
    arr = None
    if refatoms:
      if len(refatoms) != objects.size():
        raise IndigoException("renderGridToBuffer(): refatoms[] size must be equal to the number of objects")
      arr = (c_int * len(refatoms))()
      for i in xrange(len(refatoms)):
        arr[i] = refatoms[i]
    wb = self.indigo.writeBuffer()
    self.indigo._checkResult(
      self._lib.indigoRenderGrid(objects.id, arr, ncolumns, wb.id))
    return wb.toBuffer()

File: api/renderer/java/src/com/ggasoftware/indigo/IndigoRenderer.java
/****************************************************************************
 * Copyright (C) 2010-2011 GGA Software Services LLC
 *
 * This file is part of Indigo toolkit.
 *
 * This file may be distributed and/or modified under the terms of the
 * GNU General Public License version 3 as published by the Free Software
 * Foundation and appearing in the file LICENSE.GPL included in the
 * packaging of this file.
 *
 * This file is provided AS IS with NO WARRANTY OF ANY KIND, INCLUDING THE
 * WARRANTY OF DESIGN, MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
 ***************************************************************************/

package com.ggasoftware.indigo;
import java.io.*;
import com.sun.jna.*;

public class IndigoRenderer
{
   public IndigoRenderer (Indigo indigo)
   {
      loadLibrary(indigo.getUserSpecifiedPath());
      _indigo = indigo;
   }

   public void render (IndigoObject obj, IndigoObject output)
   {
      _indigo.setSessionID();
      Indigo.checkResult(this, _lib.indigoRender(obj.self, output.self));
   }

   public void renderToFile (IndigoObject obj, String filename)
   {
      _indigo.setSessionID();
      Indigo.checkResult(this, _lib.indigoRenderToFile(obj.self, filename));
   }

   public byte[] renderToBuffer (IndigoObject obj)
   {
      _indigo.setSessionID();
      IndigoObject buf = _indigo.writeBuffer();
      Indigo.checkResult(this, _lib.indigoRender(obj.self, buf.self));

      return buf.toBuffer();
   }

   public void renderGridToFile (IndigoObject objects, int[] refAtoms, int ncolumns, String filename)
   {
      _indigo.setSessionID();
      if (refAtoms != null && objects.count() != refAtoms.length)
         throw new IndigoException(this, "refAtoms size does not match the number of objects");
      Indigo.checkResult(this, _lib.indigoRenderGridToFile(objects.self, refAtoms, ncolumns, filename));
   }

   public byte[] renderGridToBuffer (IndigoObject objects, int[] refAtoms, int ncolumns)
   {
      _indigo.setSessionID();
      if (objects.count() != refAtoms.length)
         throw new IndigoException(this, "refAtoms size does not match the number of objects");
      IndigoObject buf = _indigo.writeBuffer();
      Indigo.checkResult(this, _lib.indigoRenderGrid(objects.self, refAtoms, ncolumns, buf.self));
      return buf.toBuffer();
   }

   private static String getPathToBinary (String path, String filename)
   {
      String dllpath = Indigo.getPlatformDependentPath();

      if (path == null)
      {
         String res = Indigo.extractFromJar(IndigoRenderer.class, "/com/ggasoftware/indigo/" + dllpath, filename);
         if (res != null)
            return res;
         path = "lib";
      }
      path = path + File.separator + dllpath + File.separator + filename;
      try
      {
         return (new File(path)).getCanonicalPath();
      }
      catch (IOException e)
      {
         return path;
      }
   }
   private synchronized static void loadLibrary (String path)
   {
      if (_lib != null)
         return;

      int os = Indigo.getOs();

      if (os == Indigo.OS_LINUX || os == Indigo.OS_SOLARIS)
         _lib = (IndigoRendererLib)Native.loadLibrary(getPathToBinary(path, "libindigo-renderer.so"), IndigoRendererLib.class);
      else if (os == Indigo.OS_MACOS)
         _lib = (IndigoRendererLib)Native.loadLibrary(getPathToBinary(path, "libindigo-renderer.dylib"), IndigoRendererLib.class);
      else // os == OS_WINDOWS
         _lib = (IndigoRendererLib)Native.loadLibrary(getPathToBinary(path, "indigo-renderer.dll"), IndigoRendererLib.class);
   }

   Indigo _indigo;
   static IndigoRendererLib _lib;
}
